### PR TITLE
Fix benign data race in ZooKeeper

### LINKED ISSUE #66966: Integration test `test_zookeeper_fallback_session` is flaky
https://s3.amazonaws.com/clickhouse-test-reports/0/f1296e5113653cf548f441e85f2e4cea5f7606fc/integration_tests__tsan__[1_6].html

### LINKED ISSUE #67163: Data race in ZooKeeper load balancing
https://s3.amazonaws.com/clickhouse-test-reports/0/a35a01c76c0b54d1356821fd424283ff7f52aa4c/integration_tests__tsan__[5_6]//home/ubuntu/actions-runner/_work/_temp/test/output_dir/integration_run_test_zookeeper_config_load_balancing_test_py_0.log

### PR DESCRIPTION
**Pull request by alexey-milovidov**
### Changelog category (leave one):
- Not for changelog (changelog entry is not required)

This closes #67163.
This closes #66966.

> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

#### CI Settings (Only check the boxes if you know what you are doing):
- [ ]  Allow: All Required Checks
- [ ]  Allow: Stateless tests
- [ ]  Allow: Stateful tests
- [ ]  Allow: Integration Tests
- [ ]  Allow: Performance tests
- [ ]  Allow: All Builds
- [ ]  Allow: batch 1, 2 for multi-batch jobs
- [ ]  Allow: batch 3, 4, 5, 6 for multi-batch jobs
---
- [ ]  Exclude: Style check
- [ ]  Exclude: Fast test
- [ ]  Exclude: All with ASAN
- [ ]  Exclude: All with TSAN, MSAN, UBSAN, Coverage
- [ ]  Exclude: All with aarch64, release, debug
---
- [ ]  Do not test
- [ ]  Woolen Wolfdog
- [ ]  Upload binaries for special builds
- [ ]  Disable merge-commit
- [ ]  Disable CI cache

### CODE REVIEW

**Comment by robot-ch-test-poll3** [comment]
*This is an automated comment for commit f3c88ff66707a50523ccef6e964f2fe78a711ace with description of existing statuses. It's updated for the latest CI running*

[✅ Click here](https://s3.amazonaws.com/clickhouse-test-reports/67164/f3c88ff66707a50523ccef6e964f2fe78a711ace/ci_running.html) to open a full report in a separate page

<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>Docs check</td><td>Builds and tests the documentation</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/66982/93a2dbf85c446cd5efa802dfe8fc99e2a82f394f/docs_check.html">✅ success</a></td></tr>
<tbody>
</table>
</details>

**Comment by Algunenano** [comment]
Similar PR in https://github.com/ClickHouse/ClickHouse/pull/67147, but this one looks more complete

**Comment by alexey-milovidov** [comment]
@Algunenano By the way, I also decided to remove this code initially, but then thought that it could be (barely) useful in `system.zookeeper_connection`.

File: src/Common/ZooKeeper/ZooKeeper.cpp [hunk 1]
```diff
@@ -154,13 +153,15 @@ void ZooKeeper::init(ZooKeeperArgs args_, std::unique_ptr<Coordination::IKeeper>
                     LOG_DEBUG(log, "Trying to connect to a more optimal node {}", optimal_host.host);
                     ShuffleHosts node{optimal_host};
                     std::unique_ptr<Coordination::IKeeper> new_impl = std::make_unique<Coordination::ZooKeeper>(node, args, zk_log);
-                    Int8 new_node_idx = new_impl->getConnectedNodeIdx();
 
-                    /// Maybe the node was unavailable when getting AZs first time, update just in case
-                    if (args.availability_zone_autodetect && availability_zones[new_node_idx].empty())
+                    if (auto new_node_idx = new_impl->getConnectedNodeIdx(); new_node_idx)
```
Comment (tavplubix) [inline src/Common/ZooKeeper/ZooKeeper.cpp:157]: It's always true (unless we have a bug, but it's better to add an assert or throw a LOGICAL_ERROR in this case). Moreover, even if this happens somehow, this code will ignore the issue and will finalize the current session without a reason

File: src/Common/ZooKeeper/ZooKeeperImpl.cpp [hunk 2]
```diff
@@ -1014,8 +1014,7 @@ void ZooKeeper::finalize(bool error_send, bool error_receive, const String & rea
     LOG_INFO(log, "Finalizing session {}. finalization_started: {}, queue_finished: {}, reason: '{}'",
              session_id, already_started, requests_queue.isFinished(), reason);
 
-    /// Reset the original index.
-    original_index = -1;
+    original_index.store(-1);
```
Comment (tavplubix) [inline src/Common/ZooKeeper/ZooKeeperImpl.cpp:1017]: We don't need this, see https://github.com/ClickHouse/ClickHouse/pull/67147

File: src/Storages/System/StorageSystemZooKeeperConnection.cpp [hunk 3]
```diff
@@ -27,7 +28,7 @@ ColumnsDescription StorageSystemZooKeeperConnection::getColumnsDescription()
         /* 0 */ {"name", std::make_shared<DataTypeString>(), "ZooKeeper cluster's name."},
         /* 1 */ {"host", std::make_shared<DataTypeString>(), "The hostname/IP of the ZooKeeper node that ClickHouse connected to."},
         /* 2 */ {"port", std::make_shared<DataTypeUInt16>(), "The port of the ZooKeeper node that ClickHouse connected to."},
-        /* 3 */ {"index", std::make_shared<DataTypeUInt8>(), "The index of the ZooKeeper node that ClickHouse connected to. The index is from ZooKeeper config."},
+        /* 3 */ {"index", std::make_shared<DataTypeNullable>(std::make_shared<DataTypeUInt8>()), "The index of the ZooKeeper node that ClickHouse connected to. The index is from ZooKeeper config. If not connected, this column is NULL."},
```
Comment (tavplubix) [inline src/Storages/System/StorageSystemZooKeeperConnection.cpp:31]: It still makes sense to show the node that we were connected to, there's `is_expired` to check if the connection is still alive

**Comment by alexey-milovidov** [comment]
Good, let's merge yours fix first, it looks more isolated.

**Pull request merged**